paf-mvp-core-js/src/model/audit-log.ts

```diff
@@ -0,0 +1,91 @@
+import {
+  AuditLog,
+  IdsAndPreferences,
+  Seed,
+  TransmissionRequest,
+  TransmissionResponse,
+  TransmissionResult,
+} from './generated-model';
+import { CurrentModelVersion } from './model';
+
+export const buildAuditLog = (
+  seed: Seed,
+  data: IdsAndPreferences,
+  response: TransmissionResponse,
+  contentId: string
+): AuditLog | undefined => {
+  if (response.version != CurrentModelVersion) {
+    return undefined;
+  }
+  const path = findTransactionPath(response, contentId);
+  if (path === undefined) {
+    return undefined;
+  }
+  return {
+    version: CurrentModelVersion,
+    data,
+    seed,
+    transaction_id: path.transactionId,
+    transmissions: path.results,
+  };
+};
+
+interface NodeTrack {
+  node: TransmissionResponse;
+  remainingChildren: TransmissionResponse[];
+}
+interface TransactionPath {
+  /** All the transmission results participating to the contentId, in order. */
+  results: TransmissionResult[];
+  transactionId: string;
+}
+
+const findTransactionPath = (response: TransmissionResponse, contentId: string): TransactionPath | undefined => {
+  const stack: NodeTrack[] = [];
+  let current: NodeTrack = { node: response, remainingChildren: [...response.children] };
```

Comment:
This method is quite complex and my feeling is that using one variable to hold both the current node and the remaining children is not helping.
Why not use 2 different variables?
This way I feel you could simply use `remainingChildren.length` as a test for the `while` loops. As it is today, you test if `current` is `undefined` but set it to `undefined` as soon as `remainingChildren` is empty, which feels akward.

Reply:
The method is a bit complex, indeed. However, I have good hope, for now, that is not an accidental complexity but an essential one if we consider that the recursion isn't an option, here.

I don't think it is a good idea to separate the node and the children. It would be awkward because, in a classical tree algorithm, you manipulate the current node and reach its children from it.

Reply:
I agree the complexity is not accidental but "natural", no problem with that ;-)
I would recommend adding a couple of comments then.